int-set/src/input_bit_stream.rs

```diff
@@ -1,137 +1,161 @@
 //! Reads individual bits from a array of bytes.
 
-pub(crate) struct InputBitStream<'a> {
+use crate::sparse_bit_set::BranchFactor;
+
+pub(crate) struct InputBitStream<'a, const BF: u8> {
     data: &'a [u8],
     byte_index: usize,
-    bit_index: u8,
+    sub_index: u32,
 }
 
-impl<'a> InputBitStream<'a> {
-    pub fn from(data: &'a [u8]) -> InputBitStream {
-        InputBitStream {
-            data,
-            byte_index: 0,
-            bit_index: 0,
-        }
-    }
+impl<'a, const BF: u8> Iterator for InputBitStream<'a, BF> {
+    type Item = u32;
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
This is good, but I wanted to point out for fun that in the off chance that we were really concerned about the possible branching below (which in practice shouldn't matter, the branch predictor will figure it out) we could write separate impls for the different possible values, like:

```rust
 impl<'a> Iterator for InputBitStream<'a, 2> { .. }
 impl<'a> Iterator for InputBitStream<'a, 4> { .. }
```

in any case don't do this, I just thought you might find it interesting.

Reply:
Since BF is a compile time constant my expectation is that the compiler will generate specialized versions for each branch factor where the branching is compiled out.

Reply:
hmm interesting that hadn't occurred to me but I suspect you're right!